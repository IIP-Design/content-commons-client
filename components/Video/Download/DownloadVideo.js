import React from 'react';
import PropTypes from 'prop-types';

import { formatBytes, getFileDownloadUrl } from 'lib/utils';

import DownloadItemContent from 'components/download/DownloadItem/DownloadItemContent';

// NOTE: Using the 'download' attribute to trigger downloads
// Need to research more robust options depending on browser support
const DownloadVideo = ( { burnedInCaptions, selectedLanguageUnit, isAdminPreview } ) => {
  const getSizeInfo = size => {
    if ( !size ) return null;

    return {
      label: `${size.width} x ${size.height}`,
      weight: formatBytes( size.filesize ),
    };
  };

  const getFnExt = url => {
    const extRe = /([0-9a-z]+)$/i;
    const exts = url.match( extRe );

    return exts[0];
  };

  const sortByFilesize = ( a, b ) => {
    if ( a.size && a.size.filesize && b.size && b.size.filesize ) {
      return +a.size.filesize > +b.size.filesize;
    }

    return true;
  };

  const renderFormItem = video => {
    const { title } = selectedLanguageUnit;
    const size = getSizeInfo( video.size );

    // use actual filename and not project title
    const fn = video.downloadUrl.substr( video.downloadUrl.lastIndexOf( '/' ) + 1 );

    const videoQuality = `${video.video_quality && video.video_quality.toLowerCase() === 'broadcast' ? 'broadcast' : 'web'}`;
    const downloadLink = getFileDownloadUrl( video.downloadUrl, fn );

    return (
      <DownloadItemContent
        key={ downloadLink }
        srcUrl={ downloadLink }
        hoverText={ `Download for ${videoQuality}` }
        isAdminPreview={ isAdminPreview }
        downloadFilename={ fn }
      >
        <div className="item-content">
          <p className="item-content__title">
            <strong>
              { `Download "${title}" for ${videoQuality}` }
            </strong>
          </p>
          <p className="item-content__meta">{ `${video.use} | ${video.use === 'Clean' ? 'No subtitles' : 'Subtitles'}` }</p>
          <p className="item-content__meta">{ `File size: ${size.weight}` }</p>
          <p className="item-content__meta">{ `Dimensions: ${size.label}` }</p>
        </div>
      </DownloadItemContent>
    );
  };

  const renderFormItems = ( unit, captions ) => {
    // fetch all source videos by Clean vs. non-Clean and then sort by file size
    const videos = unit.source.filter( video => {
      const isClean = video.use === 'Clean';

      return captions ? !isClean : isClean;
    } );
    const videosWithSizeProp = videos.filter( video => video.filesize );

    // only sort the videos if each video has a filesize prop for comparison
    if ( videosWithSizeProp.length === videos.length ) {
      // filesize coming in as string, convert to number for comparison
      videos.sort( sortByFilesize );
    }

    const videosArr = videos.map( ( v, i ) => v.downloadUrl && renderFormItem( v, i ) );

    return (
      <div>
        { videosArr.length
          ? videosArr
          : <p className="download-item__noContent">There are no videos available for download at this time</p> }
      </div>
    );
  };

  return (
    <div>
      { selectedLanguageUnit && renderFormItems( selectedLanguageUnit, burnedInCaptions ) }
    </div>
  );
};

DownloadVideo.propTypes = {
  selectedLanguageUnit: PropTypes.object,
  burnedInCaptions: PropTypes.bool,
  isAdminPreview: PropTypes.bool,
};

export default DownloadVideo;
