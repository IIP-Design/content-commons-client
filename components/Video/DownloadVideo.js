import React from 'react';
import PropTypes from 'prop-types';
import { Item } from 'semantic-ui-react';

import DownloadItem from './DownloadItem';
import { formatBytes, getVideoDownloadLink } from 'lib/utils';

// NOTE: Using the 'download' attribute to trigger downloads
// Need to research more robust options depending on browser support
const DownloadVideo = ( { burnedInCaptions, instructions, selectedLanguageUnit } ) => {
  const getSrt = () => {
    const arr = [];
    const unit = selectedLanguageUnit;

    if ( unit && unit.srt && unit.srt.srcUrl ) {
      arr.push( unit.srt.srcUrl );
    }

    return arr;
  };

  const getSizeInfo = size => {
    if ( !size ) return null;

    return {
      label: `${size.width} x ${size.height}`,
      weight: formatBytes( size.filesize ),
    };
  };

  const getFnExt = url => {
    const extRe = /([0-9a-z]+)$/i;
    const exts = url.match( extRe );

    return exts[0];
  };

  const sortByFilesize = ( a, b ) => {
    if ( a.size && a.size.filesize && b.size && b.size.filesize ) {
      return +a.size.filesize > +b.size.filesize;
    }

    return true;
  };

  const renderFormItem = video => {
    const { title } = selectedLanguageUnit;
    const size = getSizeInfo( video.size );
    const fn = `${title.replace( /\s/g, '_' )}_${video.size.width}.${getFnExt( video.downloadUrl )}`;
    const videoQuality = `${video.video_quality && video.video_quality.toLowerCase() === 'broadcast' ? 'broadcast' : 'web'}`;
    const downloadLink = getVideoDownloadLink( video.downloadUrl, fn );

    const header = [
      'Download ', <span key={ title } className="lightweight">{ `"${title}"` }</span>, ` for ${videoQuality}`,
    ];

    return (
      <DownloadItem
        download={ fn }
        header={ header }
        hover={ `Download for ${videoQuality}` }
        key={ video.downloadUrl }
        url={ downloadLink }
      >
        <Item.Meta>
          { ` File size: ${size.weight} ` }
        </Item.Meta>
        <Item.Meta>
          { ` Dimensions: ${size.label}` }
        </Item.Meta>
      </DownloadItem>
    );
  };

  const renderFormItems = ( unit, captions ) => {
    // fetch all source videos with NO burned in captions and then sort by file size
    const videos = unit.source.filter( video => video.burnedInCaptions === 'true' === captions );
    const videosWithSizeProp = unit.source.filter( video => video.size && video.size.filesize );

    // only sort the videos if each video has a filesize prop for comparison
    if ( videosWithSizeProp.length === videos.length ) {
      // filesize coming in as string, convert to number for comparison
      videos.sort( sortByFilesize );
    }

    const videosArr = videos.map( ( v, i ) => v.downloadUrl && renderFormItem( v, i ) );

    return <div>{ videosArr.length ? videosArr : 'There are no videos available for download at this time' }</div>;
  };

  return (
    <div>
      <div className="form-group_instructions">{ instructions }</div>
      { selectedLanguageUnit && renderFormItems( selectedLanguageUnit, burnedInCaptions ) }
    </div>
  );
};

DownloadVideo.propTypes = {
  selectedLanguageUnit: PropTypes.object,
  instructions: PropTypes.string,
  burnedInCaptions: PropTypes.bool,
};

export default DownloadVideo;
