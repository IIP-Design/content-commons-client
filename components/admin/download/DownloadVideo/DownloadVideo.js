import PropTypes from 'prop-types';

import DownloadItemContent from 'components/download/DownloadItem/DownloadItemContent';

import { formatBytes, getS3Url } from 'lib/utils';

// NOTE: Using the 'download' attribute to trigger downloads
// Need to research more robust options depending on browser support
const DownloadVideo = ( { selectedLanguageUnit, burnedInCaptions, isPreview } ) => {
  const getSizeInfo = ( dimensions, filesize ) => {
    if ( !Object.keys( dimensions ) || !filesize ) return null;

    return {
      label: `${dimensions.width} x ${dimensions.height}`,
      weight: formatBytes( filesize ),
    };
  };

  const sortByFilesize = ( a, b ) => {
    if ( a.filesize && b.filesize ) {
      return +a.filesize > +b.filesize;
    }

    return true;
  };

  const renderFormItem = video => {
    const { id, title } = selectedLanguageUnit;
    const size = getSizeInfo( video.dimensions, video.filesize );
    const videoQuality = `${video.quality && video.quality === 'BROADCAST' ? 'broadcast' : 'web'}`;

    return (
      <DownloadItemContent
        key={ `fs_${id}-${video.id}` }
        srcUrl={ getS3Url( video.url ) }
        hoverText={ `Download for ${videoQuality}` }
        isAdminPreview={ isPreview }
      >
        <div className="item-content">
          <p className="item-content__title">
            <strong>
              { `Download "${title}" for ${videoQuality}` }
            </strong>
          </p>
          <p className="item-content__meta">{ `${video.use.name} | ${video.videoBurnedInStatus === 'CLEAN' ? 'No subtitles' : 'Subtitles'}` }</p>
          <p className="item-content__meta">{ `File size: ${size.weight}` }</p>
          <p className="item-content__meta">{ `Dimensions: ${size.label}` }</p>
        </div>
      </DownloadItemContent>
    );
  };

  const renderFormItems = unit => {
    // fetch all source videos by Clean vs. non-Clean and then sort by file size
    const videos = unit.files.filter( video => {
      const isClean = video.use.name === 'Clean';

      return burnedInCaptions ? !isClean : isClean;
    } );
    const videosWithSizeProp = videos.filter( video => video.filesize );

    // only sort the videos if each video has a filesize prop for comparison
    if ( videosWithSizeProp.length === videos.length ) {
      videos.sort( sortByFilesize );
    }

    const videosArr = videosWithSizeProp
      .map( ( v, i ) => v.url && renderFormItem( v, i ) );

    return <div>{ videosArr.length ? videosArr : 'There are no videos available for download at this time' }</div>;
  };

  return selectedLanguageUnit && renderFormItems( selectedLanguageUnit, burnedInCaptions );
};

DownloadVideo.propTypes = {
  selectedLanguageUnit: PropTypes.object,
  burnedInCaptions: PropTypes.bool,
  isPreview: PropTypes.bool,
};

export default DownloadVideo;
