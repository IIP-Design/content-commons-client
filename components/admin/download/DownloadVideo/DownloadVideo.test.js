import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import DownloadVideo from './DownloadVideo';

jest.mock( 'lib/utils', () => ( {
  formatBytes: jest.fn( ( bytes, decimals ) => {
    if ( bytes === 0 ) return;
    const k = 1024;
    const dm = decimals || 2;
    const sizes = [
      'Bytes',
      'KB',
      'MB',
      'GB',
      'TB',
    ];
    const i = Math.floor( Math.log( bytes ) / Math.log( k ) );

    return `${parseFloat( ( bytes / k ** i ).toFixed( dm ) )} ${sizes[i]}`;
  } ),
  getS3Url: jest.fn( assetPath => `https://s3-url.com/${assetPath}` ),
} ) );

jest.mock( 'static/icons/icon_download.svg', () => 'downloadIconSVG' );
jest.mock( 'lib/hooks/useSignedUrl', () => jest.fn( () => ( { signedUrl: 'https://example.jpg' } ) ) );

const projectId = '123';

const props = {
  instructions: 'Download the video file in English',
  selectedLanguageUnit: {
    __typename: 'VideoUnit',
    id: 'un91',
    title: 'test project title',
    descPublic: 'the english description',
    language: {
      __typename: 'Language',
      id: 'en38',
      displayName: 'English',
      languageCode: 'en',
      locale: 'en-us',
      nativeName: 'English',
      textDirection: 'LTR',
    },
    tags: [
      {
        __typename: 'Tag',
        id: 'tag13',
        translations: [
          {
            __typename: 'LanguageTranslation',
            id: 'tr999',
            name: 'american culture',
          },
        ],
      },
    ],
    thumbnails: [
      {
        __typename: 'Thumbnail',
        id: 'th11',
        size: 'FULL',
        image: {
          __typename: 'ImageFile',
          id: 'im28',
          createdAt: '2019-03-06T13:11:48.043Z',
          updatedAt: '2019-06-18T14:58:10.024Z',
          filename: 'image-1.jpg',
          filesize: 28371,
          filetype: 'image/jpeg',
          alt: 'the alt text',
          url: `2019/06/${projectId}/image-1.jpg`,
          language: {
            __typename: 'Language',
            id: 'en38',
            displayName: 'English',
            languageCode: 'en',
            locale: 'en-us',
            nativeName: 'English',
            textDirection: 'LTR',
          },
        },
      },
    ],
    files: [
      {
        __typename: 'VideoFile',
        id: 'f19',
        createdAt: '2019-03-05T20:18:54.032Z',
        updatedAt: '2019-06-19T17:38:37.502Z',
        duration: 556,
        filename: 'video-file-1.mp4',
        filesize: 662595174,
        filetype: 'video/mp4',
        quality: 'WEB',
        url: `2019/06/${projectId}/video-file-1.mp4`,
        videoBurnedInStatus: 'CLEAN',
        dimensions: {
          __typename: 'Dimensions',
          id: 'd21',
          height: '1080',
          width: '1920',
        },
        language: {
          __typename: 'Language',
          id: 'en38',
          displayName: 'English',
          languageCode: 'en',
          locale: 'en-us',
          nativeName: 'English',
          textDirection: 'LTR',
        },
        use: {
          __typename: 'VideoUse',
          id: 'us31',
          name: 'Full Video',
        },
        stream: [
          {
            __typename: 'VideoStream',
            id: 'st93',
            site: 'YouTube',
            url: 'https://www.youtube.com/watch?v=1evw4fRu3bo',
          },
          {
            __typename: 'VideoStream',
            id: 'st35',
            site: 'Vimeo',
            url: 'https://vimeo.com/340239507',
          },
        ],
      },
      {
        __typename: 'VideoFile',
        id: 'f20',
        createdAt: '2019-03-05T20:18:54.032Z',
        updatedAt: '2019-06-19T17:38:37.502Z',
        duration: 556,
        filename: 'video-file-2.mp4',
        filesize: 652595174,
        filetype: 'video/mp4',
        quality: 'WEB',
        url: `2019/06/${projectId}/video-file-2.mp4`,
        videoBurnedInStatus: 'CLEAN',
        dimensions: {
          __typename: 'Dimensions',
          id: 'd21',
          height: '1080',
          width: '1920',
        },
        language: {
          __typename: 'Language',
          id: 'en38',
          displayName: 'English',
          languageCode: 'en',
          locale: 'en-us',
          nativeName: 'English',
          textDirection: 'LTR',
        },
        use: {
          __typename: 'VideoUse',
          id: 'us31',
          name: 'Full Video',
        },
        stream: [
          {
            __typename: 'VideoStream',
            id: 'st94',
            site: 'YouTube',
            url: 'https://www.youtube.com/watch?v=1evw4fRu3bo',
          },
          {
            __typename: 'VideoStream',
            id: 'st36',
            site: 'Vimeo',
            url: 'https://vimeo.com/340239507',
          },
        ],
      },
    ],
  },
  burnedInCaptions: false,
  isPreview: false,
};

const Component = <DownloadVideo { ...props } />;

describe( '<DownloadVideo />', () => {
  it( 'renders without crashing', () => {
    const wrapper = mount( Component );

    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders <a> tags with the correct href & download values if !isPreview', () => {
    const wrapper = mount( Component );
    const { files } = props.selectedLanguageUnit;
    const s3Bucket = 'https://s3-url.com';

    const downloadItems = wrapper.find( 'DownloadItemContent' );

    expect( downloadItems.length ).toEqual( files.length );

    downloadItems.forEach( ( item, i ) => {
      const assetPath = files[i].url;

      expect( item.prop( 'srcUrl' ) ).toEqual( `${s3Bucket}/${assetPath}` );

      const anchorLink = item.find( 'a' );

      expect( anchorLink.prop( 'href' ) ).toEqual( 'https://example.jpg' );
      expect( anchorLink.prop( 'download' ) ).toEqual( true );
    } );
  } );

  it( 'renders preview text if isPreview is true', () => {
    const wrapper = mount( Component );

    wrapper.setProps( { isPreview: true } );
    const previews = wrapper.find( '.preview-text' );
    const { files } = props.selectedLanguageUnit;

    expect( previews.length ).toEqual( files.length );
    previews.forEach( preview => {
      expect( preview.exists() )
        .toEqual( wrapper.prop( 'isPreview' ) );
      expect( preview.name() ).toEqual( 'span' );
      expect( preview.text() )
        .toEqual( 'The link will be active after publishing.' );
    } );
  } );

  it( 'does not render preview text if !isPreview', () => {
    const wrapper = mount( Component );
    const previews = wrapper.find( '.preview-text' );

    expect( previews.exists() )
      .toEqual( wrapper.prop( 'isPreview' ) );
    expect( previews.length ).toEqual( 0 );
  } );

  it( 'renders a "no video available message" if there are no video files', () => {
    const noFilesProps = {
      ...props,
      selectedLanguageUnit: {
        ...props.selectedLanguageUnit,
        files: [],
      },
    };
    const wrapper = mount( <DownloadVideo { ...noFilesProps } /> );
    const items = wrapper.find( 'a.item' );
    const { files } = noFilesProps.selectedLanguageUnit;
    const msg = 'There are no videos available for download at this time';

    expect( wrapper.exists() ).toEqual( true );
    expect( items.length ).toEqual( files.length );
    expect( wrapper.contains( msg ) ).toEqual( true );
  } );
} );
