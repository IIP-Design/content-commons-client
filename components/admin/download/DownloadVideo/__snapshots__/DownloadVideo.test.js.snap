// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadVideo />, if used for Clean videos in the "For Translations" tab renders without crashing 1`] = `
<DownloadVideo
  burnedInCaptions={false}
  instructions="Download the video file in English"
  isPreview={true}
  selectedLanguageUnit={
    Object {
      "__typename": "VideoUnit",
      "descPublic": "the english description",
      "files": Array [
        Object {
          "__typename": "VideoFile",
          "createdAt": "2019-03-05T20:18:54.032Z",
          "dimensions": Object {
            "__typename": "Dimensions",
            "height": "1080",
            "id": "d21",
            "width": "1920",
          },
          "duration": 556,
          "filename": "video-file-1.mp4",
          "filesize": 662595174,
          "filetype": "video/mp4",
          "id": "f19",
          "language": Object {
            "__typename": "Language",
            "displayName": "English",
            "id": "en38",
            "languageCode": "en",
            "locale": "en-us",
            "nativeName": "English",
            "textDirection": "LTR",
          },
          "quality": "WEB",
          "stream": Array [
            Object {
              "__typename": "VideoStream",
              "id": "st93",
              "site": "YouTube",
              "url": "https://www.youtube.com/watch?v=1evw4fRu3bo",
            },
            Object {
              "__typename": "VideoStream",
              "id": "st35",
              "site": "Vimeo",
              "url": "https://vimeo.com/340239507",
            },
          ],
          "updatedAt": "2019-06-19T17:38:37.502Z",
          "url": "2019/06/123/video-file-1.mp4",
          "use": Object {
            "__typename": "VideoUse",
            "id": "us31",
            "name": "Full Video",
          },
          "videoBurnedInStatus": "SUBTITLED",
        },
        Object {
          "__typename": "VideoFile",
          "createdAt": "2019-03-05T20:18:54.032Z",
          "dimensions": Object {
            "__typename": "Dimensions",
            "height": "1080",
            "id": "d21",
            "width": "1920",
          },
          "duration": 556,
          "filename": "video-file-2.mp4",
          "filesize": 652595174,
          "filetype": "video/mp4",
          "id": "f20",
          "language": Object {
            "__typename": "Language",
            "displayName": "English",
            "id": "en38",
            "languageCode": "en",
            "locale": "en-us",
            "nativeName": "English",
            "textDirection": "LTR",
          },
          "quality": "WEB",
          "stream": Array [
            Object {
              "__typename": "VideoStream",
              "id": "st94",
              "site": "YouTube",
              "url": "https://www.youtube.com/watch?v=1evw4fRu3bo",
            },
            Object {
              "__typename": "VideoStream",
              "id": "st36",
              "site": "Vimeo",
              "url": "https://vimeo.com/340239507",
            },
          ],
          "updatedAt": "2019-06-19T17:38:37.502Z",
          "url": "2019/06/123/video-file-2.mp4",
          "use": Object {
            "__typename": "VideoUse",
            "id": "us32",
            "name": "Clean",
          },
          "videoBurnedInStatus": "CLEAN",
        },
      ],
      "id": "un91",
      "language": Object {
        "__typename": "Language",
        "displayName": "English",
        "id": "en38",
        "languageCode": "en",
        "locale": "en-us",
        "nativeName": "English",
        "textDirection": "LTR",
      },
      "tags": Array [
        Object {
          "__typename": "Tag",
          "id": "tag13",
          "translations": Array [
            Object {
              "__typename": "LanguageTranslation",
              "id": "tr999",
              "name": "american culture",
            },
          ],
        },
      ],
      "thumbnails": Array [
        Object {
          "__typename": "Thumbnail",
          "id": "th11",
          "image": Object {
            "__typename": "ImageFile",
            "alt": "the alt text",
            "createdAt": "2019-03-06T13:11:48.043Z",
            "filename": "image-1.jpg",
            "filesize": 28371,
            "filetype": "image/jpeg",
            "id": "im28",
            "language": Object {
              "__typename": "Language",
              "displayName": "English",
              "id": "en38",
              "languageCode": "en",
              "locale": "en-us",
              "nativeName": "English",
              "textDirection": "LTR",
            },
            "updatedAt": "2019-06-18T14:58:10.024Z",
            "url": "2019/06/123/image-1.jpg",
          },
          "size": "FULL",
        },
      ],
      "title": "test project title",
    }
  }
>
  <div>
    <DownloadItemContent
      hoverText="Download for web"
      isAdminPreview={true}
      key="fs_un91-f20"
      srcUrl="https://s3-url.com/2019/06/123/video-file-2.mp4"
    >
      <a
        className="download-item download-item--preview"
        download={true}
        href="https://example.jpg"
        key="https://example.jpg"
        rel="noopener noreferrer"
      >
        <div
          className="item-icon"
        >
          <img
            alt="Download for web"
            src="downloadIconSVG"
          />
        </div>
        <div
          className="item-content"
        >
          <p
            className="item-content__title"
          >
            <strong>
              Download "test project title" for web
            </strong>
          </p>
          <p
            className="item-content__meta"
          >
            Clean | No subtitles
          </p>
          <p
            className="item-content__meta"
          >
            File size: 622.36 MB
          </p>
          <p
            className="item-content__meta"
          >
            Dimensions: 1920 x 1080
          </p>
        </div>
        <span
          className="item-hover"
        >
          Download for web
          <span
            className="preview-text"
          >
            The link will be active after publishing.
          </span>
        </span>
      </a>
    </DownloadItemContent>
  </div>
</DownloadVideo>
`;
